052-cruise-control-api-users.md

```diff
@@ -0,0 +1,261 @@
+It also prevents developers from enabling third party applications such as the Cruise Control UI which could make monitoring their Cruise Control data easier.
+
+In order to allow developers and third-party applications to access the Cruise Control API without needing to disable API security, we would need to provide a method of creating API users.
+
+## Proposal
+
+### Goals
+
+* A method of creating Cruise Control API users.
+
+### Developer managed API user configuration
+
+Having the developer manage a single secret with their custom Cruise Control API user configuration would greatly reduce the complexity of the proposed design.
+
+Cruise Control's auth credential files follow Jetty's HashLoginService's file format: `username: password [,rolename ...]`.
+
+By default Cruise Control defines three roles: `VIEWER`, `USER` and `ADMIN`.
+```
+VIEWER role: has access to the most lightweight kafka_cluster_state, user_tasks and review_board endpoints.
+USER role: has access to all the GET endpoints except bootstrap and train.
+ADMIN role: has access to all endpoints.
+```
+For more details checkout the [Cruise Control Wiki](https://github.com/linkedin/cruise-control/wiki/Security#authorization)
+
+So a developer could create a Cruise Control API auth credentials file, `cruise-control-auth.txt`, that looked like this:
+```
+userOne: passwordOne, USER
+userTwo: passwordTwo, VIEWER
+```
+
+Then the developer could create a secret using that file:
+
+```
+kubectl create secret generic cruise-control-api-users-secret  --from-file=key=cruise-control-auth.txt
+```
+
+This would create a Secret:
+
+```yaml=
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cruise-control-api-users-secret
+type: Opaque
+data:
+  key: dXNlck9uZTogcGFzc3dvcmRPbmUsIFVTRVIKdXNlclR3bzogcGFzc3dvcmRUd28sIFZJRVdFUgo=
+```
+
+The developer could then reference the secret in an `spec.cruiseControl.apiUsers` section of the `Kafka` resource.
+
+The schema would look like this:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  ...
+spec:
+  cruiseControl:
+    apiUsers:
+      type: fromSecret (1)
+      secret:
+        secretName: cruise-control-api-users-secret
+     ...
+```
+(1) A type field is added here for configuration flexibility in the future.
+
+Strimzi would decode and use the contents of this secret to populate the Cruise Control API auth credentials file which would be mounted and used by the Cruise Control pod.
+
+Although this would simplify the overall design it would also shift a lot of the maintanence responsibility to the developer.
+The developer would need to create a valid Cruise Control API user configuration and encode it into a secret.
+That being said, this is still a reasonable amount of responsibility to leave to the developer.
+This is because accessing the Cruise Control API directly is neither the supported or primary way of interacting with Cruise Control for a Strimzi-managed Kafka cluster.
```

Comment:
Maybe like this, it would sound better? _Supported_ is a bit confusing term in OSS project. We do not provide any support as a _helpline_. Everything what the code allows is basically supported and unsupported in the same time.

```suggestion
This is because accessing the Cruise Control API directly is not the primary way of interacting with Cruise Control for a Strimzi-managed Kafka cluster and is expected to be used only by advanced users in special use-cases.
```

Reply:
I don't know if we want to say anything here about the need to document the approach and underlining in the documentation that this is for advanced Cruise Control operations?

Something like: "However, with proper supporting documentation, this new feature will give developers a greater degree of control and assistance with certain use-cases."